cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -523,18 +539,37 @@ class TimestampConverter
   TimeUnit::type unit_;
 };
 
+template <bool from_pandas = true>
 class Float16Converter
-    : public TypedConverterVisitor<HalfFloatBuilder, Float16Converter> {
+    : public TypedConverterVisitor<HalfFloatBuilder, Float16Converter<from_pandas>> {
  public:
   // Append a non-missing item
   Status AppendItem(PyObject* obj) {
     npy_half val;
     RETURN_NOT_OK(PyFloat_AsHalf(obj, &val));
-    return typed_builder_->Append(val);
+    return this->typed_builder_->Append(val);
+  }
+};
+
+template <>
+class Float16Converter<false>
+    : public TypedConverterVisitor<HalfFloatBuilder, Float16Converter<false>> {
+ public:
+  Status AppendSingle(PyObject* obj) override {
```

Comment:
It's a bit of a pity that we're repeating a lot of code here (both `AppendSingle` and the type-specific `AppendItem` implementations). C++ doesn't help us, but is there a simple way to avoid that repetition (perhaps inheriting some mixins)?

Reply:
Quite possible, I will look more closely and see if any ideas

Reply:
Tried to do this at first but did not find a solution to this. Then I revert to using these more verbose class specifications.